### ISSUE TITLE
List Navigation Reorder Parents

### ISSUE THREAD

**Issue by jrbostic** [superlou/fountain-atom#28]
When a parent node is moved in the outline view, child nodes/scenes should be moved also.  This is an improvement on issue #19.

**Comment by jrbostic** [comment]
I've been looking at this tonight, and we may need to have a discussion about the requirements of this feature.  It's complicated enough that I don't want to go into it with assumptions about expected behavior and find out later that it doesn't fit the group vision.

These are the basic rules that I'm seeing...
1. items are able to move to the same depth under any parent (or no parent) across the document
2. parents retain children, but must remain subject to rule #1 
3. no item's depth in the hierarchy shall ever be modified by outline view

Anything wrong or missing?

**Comment by superlou** [comment]
So, I'm not sure if it's that simple or not.  Fountain's kind-of a weird syntax to think about hierarchically, since scenes are not hierarchical.  There's no such thing as a level 2 or a level 3 scene.  I think it's more about trying to guess what the dragger's intent is.

Dragging a scene acts on everything in the scene up to the next scene, section, or transition (anything other than dialogue or action).  Then when the drop ends, it gets a new level based on wherever you dropped it, since scenes don't really have a level of their on.

Since you can't know if the dragger wants you to move a section "in" or "out" in the hierarchy, what if dragging a section is relatively simple: it moves everything in the section, keeping the "level" since the section identifier isn't changed.  Then it's up to the user to fix the section identifier if they intended it to change.

I've used UIs like Wordpress's Menu Editor that let you explicitly set the new hierarchy as you drag up and down by letting you shift the item left and right during the drag.  Something like [nestable](https://dbushell.com/Nestable/).

**Comment by jrbostic** [comment]
Yea, I remember looking at nestable while working on this, but (as the author mentioned) it's incomplete in many ways.  At the very least, we need a proper callback hook for detecting list changes... (If I remember right, that may have been why I decided against nestable).  But I agree that this would be a pretty cool feature.

Part of the way I was able to make sorting work with the current library is by flattening out the data structure and moving the hierarchical indentation to css.  The original hierarchical data structure is still available, but it's not the actual model backing the view anymore.

The information we currently have about an element on drag are its index, start line, type, and depth.  

And now that I think about it... I'm a little concerned that drag/drop may not work correctly when hiding scenes... And I just confirmed that this is true.  Crap.  So,  I guess you probably shouldn't do a release until that issue is resolved.  It seems the simplest tentative fix would be just to disable reordering when scenes are hidden.  A solution for the parent dragging could then be hooked into the hidden scene state.  I'm sorry for missing that.

**Comment by jrbostic** [comment]
I'll have a PR up by tonight to "fix" the above-mentioned bug... In case you'd like to do a release before all of this parent reordering stuff is completed.

**Comment by superlou** [comment]
I think I'd like the parent-reordering behavior in before a release so that users don't have two different behaviors to learn.

**Comment by jrbostic** [comment]
Okay, sounds good.  Disregard PR then.  Also, thanks for the feedback about structure.  Maybe I'm imagining problems that won't actually come to fruition.  Usually, I have to take a step back and reevaluate my assumptions about a problem like this.  Maybe it will become even simpler after solving the parent problem.  We'll see what comes of it.

**Comment by jrbostic** [comment]
Hey, so I was wondering if you might be willing to consider an idea?  I think it would be possible to comment out the lock icon in the outline view and disallow access to scene reordering... And this might allow for a minor release.  Wondering if you would be open to that while I work on the parent reordering?  My concern is holding up feedback and then potentially getting several issues at once (related to both features) that I can't find time to help with.  And while I'm getting closer to a solution on parent reordering, there's also the related tasks of generating a new test file and writing the more difficult tests... Hopefully in a more maintainable way.

**Comment by superlou** [comment]
Just to make sure I understand: you're suggesting we disable (leaving the code in, though) all drag-drop reordering so that we can make a release of the PDF feature?  If that's the case, I'm on board.

**Comment by jrbostic** [comment]
That is the case.  I'm thinking that it would take commenting out or modifying like... 2-3 lines?  Then the lock would not be visible.

**Comment by jrbostic** [comment]
And there are no other hooks to locking and unlocking at current.

**Comment by superlou** [comment]
That sounds fine.  If you make a PR, I'll merge then release.

**Comment by superlou** [comment]
Welp, master should be ready to publish as 0.6.0.  Unfortunately, my APM/git got messed up.  I use 2 factor auth on Github, and APM isn't accepting my token, so the push of the latest version failed.  Now I can't publish because according to NPM, 0.6.0 exists, and I can't APM unpublish because apparently 0.6.0 is not found.

**Comment by jrbostic** [comment]
So, I got this mostly working a couple days ago, however in the process I noticed a bug in what I believe to be the watcher for editor changes... that I'm still tracking down.  It's kind of like the others that I fixed a while back, in that it probably is causing atom to eventually become bogged down if the outline view is getting used on multiple files over a long period with no restart.  Even though most people are probably not seeing it... I want to knock it out, since I've added additional recursive processing that can cause the heap to run out of memory faster if not cleaned up.

1.  Fix call stacking bug
2.  Ensure headings and children are dropped after siblings and children when scenes are hidden (as visual logic dictates), instead of within the sibling as a child
3.  Reapply scene hiding after drag and drop
4.  Update tests

I'm also thinking to add an issue for using Ctrl+z from the outline view so you can immediately undo changes made via outline, without bringing the fountain file into focus.  I may just do that in this branch if it's easy enough.

Probably a week or two out.

**Comment by jrbostic** [comment]
I've been busy, so this got temporarily pushed to the back of my mind.  Branch still in progress.  Sorry for delay.

**Comment by jrbostic** [comment]
Alright, so going over the checklist above...

1. Couldn't produce this again.  Putting it off, as I think it would just involve a refactor to use promises instead of traditional recursion.  Will be easier with better tests too.
2. Done.
3. Done.
4. Pending.  I'll be creating a simplified test file for this, so as to assist debugging in the future, then I'll rewrite the tests as systematically as I can.

Feeling close the the finish line.  Happy with behavioral predictability.  I just need tests that are easier to understand and maintain...

**Comment by superlou** [comment]
Thanks for the update!  I [can't update the package anyway](https://discuss.atom.io/t/apm-not-recognizing-login-unable-to-publish/50526).

**Comment by jrbostic** [comment]
So, I'm pretty confident at this point.  Fixed some additional bugs that I came across manually, updated existing tests, and added a simplified test file.

Unfortunately, my hope of writing awesome/extensive/intuitive tests for this may not ever come to fruition.  The ideal, I think, would be to interact with the drag/drop events on the outline list... but the support that I found for that is limited and outdated (for unit testing).  Another way might be to set up some kind of selenium webdriver e2e thing.  I just can't commit to hacking away at that right now.  I guess the downside of all of this is that the tests will be somewhat brittle.... sensitive to changes in the underlying data structures.  And costly... in terms of time now and later.

I'm going to put up a pull request now.

**Comment by jrbostic** [comment]
PR #44

**Comment by jrbostic** [comment]
I was able to get the hardest part of some improved automated testing infrastructure added.  Traded off redundancy and test complexity for some additional configuration complexity.  Now I know that I dreaded it with good reason.  I added a few more tests... but hopefully it will prove useful in the future as well.  I also manually tested quite a bit, but if anyone could check it out and run it through some scenarios, I'd appreciate it.
